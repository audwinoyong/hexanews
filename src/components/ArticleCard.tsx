import * as React from 'react'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import CardMedia from '@material-ui/core/CardMedia'
import Typography from '@material-ui/core/Typography'
import { History } from 'history'

import placeholder from '../assets/placeholder.png'

type Props = {
  title: string,
  description: string,
  history: History,
}

const styles = {
  root: {
    flexGrow: 1,
    cursor: 'pointer'
  },
  card: {
    maxWidth: 385,
    boxShadow: '0px 0.25px 0.35px 0.35px rgba(0, 0, 0, 0.1)'
  },
  media: {
    height: 0,
    paddingTop: '56.25%',
  },
}

const ArticleCard: React.StatelessComponent<Props> = ({
  history,
  title,
  description
}) => {
  return (
    <div style={styles.root} onClick={() => history.push('/postDetail')}>
      <Card style={styles.card}>
        <CardMedia
          style={styles.media}
          image={placeholder}
        />
        <CardContent>
          <Typography
            component='h2'
            gutterBottom
            variant='headline'
          >
            {title}
          </Typography>
          <Typography component='p'>
            {description}
          </Typography>
        </CardContent>
      </Card>
    </div>
  )
}

export default ArticleCard