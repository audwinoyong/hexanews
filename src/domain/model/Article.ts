/**
 * Model for the article.
 */
export type Article = {
  title: string,
  description: string
  category?: string
  authorId?: string
}