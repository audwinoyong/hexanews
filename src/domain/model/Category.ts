/**
 * Model for category object in our application.
 */

export type Category = {
  _id?: string,
  title: string,
  description: string,
}