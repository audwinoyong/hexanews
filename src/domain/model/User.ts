/**
 * Model for the user.
 */
export type User = {
  _id: string,
  email: string,
  name: string,
  description: string,
  createdAt: string,
}