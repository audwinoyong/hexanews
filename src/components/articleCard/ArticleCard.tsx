/**
 * Display article card component.
 */

import * as React from 'react'
import { Dispatch } from 'redux'
import { push } from 'connected-react-router'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import CardMedia from '@material-ui/core/CardMedia'
import Typography from '@material-ui/core/Typography'
import * as moment from 'moment'

import placeholder from '../../assets/placeholder.png'
import styles from './styles'
import * as Config from '../../constants/config'
import { Article } from '../../domain/model/Article'
import Utils from '../../utils'

type Props = {
  article: Article,
  authorName: string,
  dispatch: Dispatch<any>,
}

const ArticleCard: React.StatelessComponent<Props> = ({
  article,
  authorName,
  dispatch,
}) => {
  const articleImage = Utils.getFeaturedImage(article) ? Utils.getFeaturedImage(article) : placeholder
  return (
    <div id={`article-${article.title}`}
      style={styles.root}
      onClick={() => dispatch(push(`${Config.HEADER_LINK.articleDetail}/${article._id}`))}
    >
      <Card style={styles.card}>
        <CardMedia
          style={styles.media}
          image={articleImage}
        />
        <CardContent>
          <Typography
            component='h3'
            gutterBottom
            variant='headline'
            style={{ fontWeight: 'bold' }}
          >
            {article.title}
          </Typography>
          <Typography
            variant='body1'
            style={styles.author}
          >
            {authorName}
          </Typography>
          <Typography
            variant='body1'
            color='textSecondary'
            style={styles.date}
          >
            {moment(article.createdAt).format('D MMMM YYYY')}
          </Typography>
        </CardContent>
      </Card>
    </div>
  )
}

export default ArticleCard