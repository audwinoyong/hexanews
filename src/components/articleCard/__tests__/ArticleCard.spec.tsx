import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

import ArticleCard from '../ArticleCard'
import { Article } from '../../../domain/model/Article'

Enzyme.configure({ adapter: new Adapter() })

const mockArticle: Article = {
  _id: '2',
  title: 'def',
  description: 'tester',
  category: 'tester',
  authorId: 'tester',
}

describe('Article Card', () => {
  it('renders Article Card', () => {
    const component = renderer.create(
      <ArticleCard
        article={mockArticle}
        dispatch={jest.fn()}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle button go to article detail', () => {
    const dispatch = jest.fn()
    const component = (
      <ArticleCard
        article={mockArticle}
        dispatch={dispatch}
      />
    )
    const layout = shallow(component)
    layout.find('#article-test').first().simulate('click')
    expect(dispatch).toHaveBeenCalled()
  })
})