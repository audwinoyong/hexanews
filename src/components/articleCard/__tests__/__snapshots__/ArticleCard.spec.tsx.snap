// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Card renders Article Card 1`] = `
<div
  id="article-def"
  onClick={[Function]}
  style={
    Object {
      "cursor": "pointer",
      "flexGrow": 1,
    }
  }
>
  <div
    className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
    style={
      Object {
        "boxShadow": "0px 0.25px 0.35px 0.35px rgba(0, 0, 0, 0.1)",
        "maxWidth": 385,
      }
    }
  >
    <div
      className="MuiCardMedia-root-29"
      src={undefined}
      style={
        Object {
          "height": 0,
          "paddingTop": "56.25%",
        }
      }
    />
    <div
      className="MuiCardContent-root-31"
    >
      <h3
        className="MuiTypography-root-32 MuiTypography-headline-37 MuiTypography-gutterBottom-49"
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        def
      </h3>
      <p
        className="MuiTypography-root-32 MuiTypography-body1-41"
        style={
          Object {
            "marginTop": 14,
          }
        }
      >
        tim
      </p>
      <p
        className="MuiTypography-root-32 MuiTypography-body1-41 MuiTypography-colorTextSecondary-55"
        style={
          Object {
            "marginTop": 2,
          }
        }
      >
        20 September 2018
      </p>
    </div>
  </div>
</div>
`;
